### ISSUE TITLE
implement (or find) a ResNet for 3D volume to single scalar

### ISSUE THREAD

**Issue by alisterburt** [warpem/miss-alignment#5]

**Comment by McHaillet** [comment]
I will update this list if I find something applicable

- [ ] ~~torch vision includes a 3D ResNet model -- https://pytorch.org/vision/0.8/models.html  (specific for video)~~
- [ ] ~~this is voxresnet but its for segmentation, so not to single scalar -- https://github.com/bo-10000/pytorch_3d_segmentation/blob/master/voxresnet.py~~
- [ ] this repo has a few classification models for 3D inputs: https://github.com/ZFTurbo/classification_models_3D/tree/master
- [ ] monai has option for residual blocks with varying spatial dimensions, could be useful to built a network: https://docs.monai.io/en/0.3.0/networks.html#residualunit
- [ ] monai also has a ResNet model with varying spatial dimensions; probably a good starting point: https://docs.monai.io/en/stable/networks.html#resnet

(monai version was adapted from this repo: https://github.com/kenshohara/3D-ResNets-PyTorch/tree/master/models)

**Comment by alisterburt** [comment]
minor point but their implementation seems more focused on images over time than 3D images so may be more complex than required

it would be great if you used https://lightning.ai/docs/pytorch/stable/ for this model - it basically provides a structure in which to implement your model and simplifies experimentation after setup, I used it for https://github.com/teamtomo/fidder and it was a good experience 🙂

**Comment by McHaillet** [comment]
Ah right, its specifically for time series of images

Sounds good, I know Ilja was always a big fan of pytorch-lightning. I'll check it out!

**Comment by McHaillet** [comment]
Just putting this there for layout inspiration. This one is for MRI, hence the strange dimension changes. And for multiple class predictions. 

![Image](https://github.com/user-attachments/assets/763ede7a-3e9e-4520-b385-ede390095d8f)

**Comment by alisterburt** [comment]
Nice - looks good, here is what Claude gave me in one shot when asked for a 3D ResNet which outputs a single scalar from 128^3 volumes

```python

import torch
import torch.nn as nn
import torch.nn.functional as F


class BasicBlock3D(nn.Module):
    expansion = 1

    def __init__(self, in_channels, out_channels, stride=1, downsample=None):
        super(BasicBlock3D, self).__init__()
        self.conv1 = nn.Conv3d(in_channels, out_channels, kernel_size=3, 
                               stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm3d(out_channels)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = nn.Conv3d(out_channels, out_channels, kernel_size=3,
                               stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm3d(out_channels)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out


class Bottleneck3D(nn.Module):
    expansion = 4

    def __init__(self, in_channels, out_channels, stride=1, downsample=None):
        super(Bottleneck3D, self).__init__()
        self.conv1 = nn.Conv3d(in_channels, out_channels, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm3d(out_channels)
        self.conv2 = nn.Conv3d(out_channels, out_channels, kernel_size=3, stride=stride,
                               padding=1, bias=False)
        self.bn2 = nn.BatchNorm3d(out_channels)
        self.conv3 = nn.Conv3d(out_channels, out_channels * self.expansion, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm3d(out_channels * self.expansion)
        self.relu = nn.ReLU(inplace=True)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out


class ResNet3D(nn.Module):
    def __init__(self, block, layers, num_input_channels=1, num_classes=1):
        super(ResNet3D, self).__init__()
        self.in_channels = 64
        
        # Initial convolution layer
        self.conv1 = nn.Conv3d(num_input_channels, 64, kernel_size=7, stride=2, padding=3, bias=False)
        self.bn1 = nn.BatchNorm3d(64)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool3d(kernel_size=3, stride=2, padding=1)
        
        # Residual layers
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2)
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2)
        self.layer4 = self._make_layer(block, 512, layers[3], stride=2)
        
        # Global average pooling and fully connected layer
        self.avgpool = nn.AdaptiveAvgPool3d((1, 1, 1))
        self.fc = nn.Linear(512 * block.expansion, num_classes)
        
        # Initialize weights
        for m in self.modules():
            if isinstance(m, nn.Conv3d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
            elif isinstance(m, nn.BatchNorm3d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

    def _make_layer(self, block, out_channels, blocks, stride=1):
        downsample = None
        if stride != 1 or self.in_channels != out_channels * block.expansion:
            downsample = nn.Sequential(
                nn.Conv3d(self.in_channels, out_channels * block.expansion,
                          kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm3d(out_channels * block.expansion),
            )

        layers = []
        layers.append(block(self.in_channels, out_channels, stride, downsample))
        self.in_channels = out_channels * block.expansion
        for _ in range(1, blocks):
            layers.append(block(self.in_channels, out_channels))

        return nn.Sequential(*layers)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        x = self.fc(x)

        return x


def resnet3d_18(num_input_channels=1, num_classes=1):
    """
    Constructs a ResNet-18 model with 3D convolutions
    """
    return ResNet3D(BasicBlock3D, [2, 2, 2, 2], 
                    num_input_channels=num_input_channels, 
                    num_classes=num_classes)


def resnet3d_34(num_input_channels=1, num_classes=1):
    """
    Constructs a ResNet-34 model with 3D convolutions
    """
    return ResNet3D(BasicBlock3D, [3, 4, 6, 3], 
                    num_input_channels=num_input_channels, 
                    num_classes=num_classes)


def resnet3d_50(num_input_channels=1, num_classes=1):
    """
    Constructs a ResNet-50 model with 3D convolutions
    """
    return ResNet3D(Bottleneck3D, [3, 4, 6, 3], 
                    num_input_channels=num_input_channels, 
                    num_classes=num_classes)


def resnet3d_101(num_input_channels=1, num_classes=1):
    """
    Constructs a ResNet-101 model with 3D convolutions
    """
    return ResNet3D(Bottleneck3D, [3, 4, 23, 3], 
                    num_input_channels=num_input_channels, 
                    num_classes=num_classes)


# Example usage
if __name__ == "__main__":
    # Create a sample 3D volume of size 128x128x128 with 1 channel
    # Batch size of 2
    sample_input = torch.randn(2, 1, 128, 128, 128)
    
    # Create a ResNet-18 model
    model = resnet3d_18(num_input_channels=1, num_classes=1)
    
    # Forward pass
    output = model(sample_input)
    
    print(f"Input shape: {sample_input.shape}")
    print(f"Output shape: {output.shape}")
    print(f"Model parameters: {sum(p.numel() for p in model.parameters() if p.requires_grad)}")
```

I'm not sure how much you've played with these things before, if torch abstractions are all a bit new I'd recommend 
- stepping through the forward pass in the debugger so you can see how data flows 
- try to put it into an optimization loop so you can start seeing how gradients flow

Ping if you need any input!

Also not sure how much you're using Claude/ChatGPT these days - Dimitry and I are both heavy Claude users

**Comment by McHaillet** [comment]
:eyes:

Thanks for sharing. Did not expect it to produce it so well... But I guess there are many training examples of torch models online!

> stepping through the forward pass in the debugger so you can see how data flows

Will give that a go

**Comment by dtegunov** [comment]
Sorry, I should have clarified: the architecture of a 3D ResNet is trivial at this point, and most LLMs will probably 0-shot it. The intent behind finding one that already exists is that it might already be pretrained, thus not requiring you to train the low-level kernels from scratch.

A model big enough to classify many complex features may also be overkill here. I'd speculate the features indicating good alignment may be much simpler, requiring fewer parameters. I'd start with something shallow and narrow, and make sure it sees a lot of augmented training data since even few parameters can still overfit.

**Comment by McHaillet** [comment]
> The intent behind finding one that already exists is that it might already be pretrained, thus not requiring you to train the low-level kernels from scratch.

That's an interesting point. I think for now I will start with the basic implementation and train from scratch, but perhaps later on the lower few layers of some medical 3d network might worthwhile (a quick search found me this: https://huggingface.co/TencentMedicalNet/MedicalNet-Resnet50).

For network size I was thinking of just starting with the 18 layer resnet (also easiest for local testing). But I guess that is already a fairly small network; reducing the size too much also removes some of the abstractions it might make.
